nitro-protocol/contracts/VirtualMarginApp.sol

```diff
@@ -0,0 +1,146 @@
+            require(candidate.variablePart.turnNum == 2, 'bad candidate turnNum; |proof|=1');
+            require(
+                NitroUtils.isClaimedSignedBy(candidate.signedBy, 2),
+                'redemption not signed by Bob'
+            );
+            MarginState memory receipt = requireValidMarginState(candidate.variablePart.appData, fixedPart);
+            requireCorrectAdjustments(
+                proof[0].variablePart.outcome,
+                candidate.variablePart.outcome,
+                receipt
+            );
+            return;
+        }
+        revert('bad proof length');
+    }
+
+    function requireProofOfUnanimousConsensusOnPostFund(
+        RecoveredVariablePart memory rVP,
+        uint256 numParticipants
+    ) internal pure {
+        require(rVP.variablePart.turnNum == 1, 'bad proof[0].turnNum; |proof|=1');
+        require(
+            NitroUtils.getClaimedSignersNum(rVP.signedBy) == numParticipants,
+            'postfund !unanimous; |proof|=1'
+        );
+    }
+
+    function requireValidMarginState(bytes memory appData, FixedPart memory fixedPart)
+        internal
+        pure
+        returns (MarginState memory)
+    {
+        MarginState memory receipt = abi.decode(appData, (MarginState));
+        bytes32 receiptHash = keccak256(abi.encode(NitroUtils.getChannelId(fixedPart), receipt.leaderAmount, receipt.followerAmount, receipt.version));
+
+        address firstSigner = NitroUtils.recoverSigner(
+            receiptHash,
+            receipt.leaderSig
+        );
+        address lastSigner = NitroUtils.recoverSigner(
+            receiptHash,
+            receipt.followerSig
+        );
+        require(firstSigner == fixedPart.participants[0], 'invalid signature from Leader');
+        require(lastSigner == fixedPart.participants[fixedPart.participants.length - 1], 'invalid signature from Follower');
+        return receipt;
+    }
+
+    function requireCorrectAdjustments(
+        Outcome.SingleAssetExit[] memory oldOutcome,
+        Outcome.SingleAssetExit[] memory newOutcome,
+        MarginState memory receipt
+    ) internal pure {
+        // TODO: Validate collateral asset type is valid
+        //
+        // require(
+        //     oldOutcome.length == 1 &&
+        //         newOutcome.length == 1 &&
+        //         oldOutcome[0].asset == address(0) &&
+        //         newOutcome[0].asset == address(0),
+        //     'only native asset allowed'
+        // );
+
+        require(
+            newOutcome[0].allocations[uint256(AllocationIndices.Leader)].amount == receipt.leaderAmount,
+            'Leader not adjusted correctly'
+        );
+        require(
+            newOutcome[0].allocations[uint256(AllocationIndices.Follower)].amount == receipt.followerAmount,
+            'Follower not adjusted correctly'
+        );
+        // TODO: Find a way to validate version number
```

Comment:
There is no way to store margin version number in VirtualMarginApp as `requireStateSupported` must be `pure`.